agent/config/builder.go

```diff
@@ -185,11 +190,14 @@ func (b *Builder) ReadPath(path string) ([]Source, error) {
 			continue
 		}
 
-		src, err := b.ReadFile(fp)
-		if err != nil {
-			return nil, err
+		sourceFormat := FormatFrom(fp)
+		if sourceFormat == "json" || sourceFormat == "hcl" || configFormat != "" {
```

Comment:
Would that read a config file where `sourceFormat: hcl` with `configFormat: json`? I am wondering if that could be a counter intuitive.

Reply:
Would it read the file when sourceFormat: "" and configFormat: json? On of the issues mentioned that it reads something.json~ which was unexpected. I think this can still happen?

Reply:
It would read the config in both of those cases.

- If `sourceFormat == hcl` and `configFormat == json` then the file is interpreted as JSON. 
- If `sourceFormat == ""` and `configFormat == json` then the file is also interpreted as JSON.

These are subtly different from the issue where someone was reading `something.json~`. In that case, the user was only setting the `config-dir` and not the `config-format`. So after this PR their file would not be read because `sourceFormat == "json" || sourceFormat == "hcl" || configFormat != ""` is false (sourceFormat comes from the file suffix).

So of those two examples I think the behavior of the 2nd one is OK. 

In the 1st case I agree that having the hcl/json mismatch would be odd. However that's the behavior stated for [config-format](https://www.consul.io/docs/agent/options.html#_config_format)

The CLI is also explicit about the config-format flag:
> Config files are in this format irrespective of their extension

Maybe this is a case where we could throw a warning if that HCL/JSON mismatch is spotted.

Reply:
I see! Thanks for explaining!